flink-runtime/src/test/java/org/apache/flink/runtime/operators/hash/ReduceHashTableTest.java

```diff
@@ -0,0 +1,501 @@
+				Tuple2<Long, String> next;
+				while ((next = iter.next()) != null) {
+					assertNotNull(next.f0);
+					assertNotNull(next.f1);
+					assertEquals(next.f0.longValue(), Long.parseLong(next.f1));
+
+					bitSet.set(next.f0.intValue());
+				}
+
+				assertEquals(numElements, bitSet.cardinality());
+			}
+
+			// make sure all entries are contained via the prober
+			{
+				ReduceHashTable<Tuple2<Long, String>>.HashTableProber<Long> proper =
+					table.getProber(probeComparator, pairComparator);
+
+				Tuple2<Long, String> reuse = new Tuple2<>();
+
+				for (long i = 0; i < numElements; i++) {
+					assertNotNull(proper.getMatchFor(i, reuse));
+					assertNull(proper.getMatchFor(i + numElements, reuse));
+				}
+			}
+		}
+		catch (Exception e) {
+			e.printStackTrace();
+			fail(e.getMessage());
+		}
+	}
+
+	/**
+	 * This test validates that records are not lost via "insertOrReplace()" as in bug [FLINK-2361]
+	 */
+	@Test
+	public void testHashTableGrowthWithInsertOrReplace() {
+		try {
+			final int numElements = 1000000;
+
+			List<MemorySegment> memory = getMemory(1000, 32 * 1024);
+
+			// we create a hash table that thinks the records are super large. that makes it choose initially
+			// a lot of memory for the partition buffers, and start with a smaller hash table. that way
+			// we trigger a hash table growth early.
+			ReduceHashTable<Tuple2<Long, String>> table = new ReduceHashTable<Tuple2<Long, String>>(
+				serializer, comparator, memory, null, null, false);
+			table.open();
+
+			for (long i = 0; i < numElements; i++) {
+				table.insertOrReplaceRecord(new Tuple2<Long, String>(i, String.valueOf(i)));
+			}
+
+			// make sure that all elements are contained via the entry iterator
+			{
+				BitSet bitSet = new BitSet(numElements);
+				MutableObjectIterator<Tuple2<Long, String>> iter = table.getEntryIterator();
+				Tuple2<Long, String> next;
+				while ((next = iter.next()) != null) {
+					assertNotNull(next.f0);
+					assertNotNull(next.f1);
+					assertEquals(next.f0.longValue(), Long.parseLong(next.f1));
+
+					bitSet.set(next.f0.intValue());
+				}
+
+				assertEquals(numElements, bitSet.cardinality());
+			}
+
+			// make sure all entries are contained via the prober
+			{
+				ReduceHashTable<Tuple2<Long, String>>.HashTableProber<Long> proper =
+					table.getProber(probeComparator, pairComparator);
+
+				Tuple2<Long, String> reuse = new Tuple2<>();
+
+				for (long i = 0; i < numElements; i++) {
+					assertNotNull(proper.getMatchFor(i, reuse));
+					assertNull(proper.getMatchFor(i + numElements, reuse));
+				}
+			}
+
+		}
+		catch (Exception e) {
+			e.printStackTrace();
+			fail(e.getMessage());
+		}
+	}
+
+	/**
+	 * This test validates that new inserts (rather than updates) in "insertOrReplace()" properly
+	 * react to out of memory conditions.
+	 *
+	 * Btw. this also tests the situation, when records are larger than one memory segment.
+	 */
+	@Test
+	public void testInsertsWithInsertOrReplace() {
```

Comment:
Good catch!

Btw. this test is basically a copy-paste from `CompactingHashTableTest.testInsertsWithInsertOrReplace`, so I will fix this there too.